crates/trigger/src/runtime_config/variables_provider.rs

```diff
@@ -79,35 +83,52 @@ pub struct VaultVariablesProviderOpts {
 }
 
 impl VaultVariablesProviderOpts {
-    pub fn build_provider(&self) -> VariablesProvider {
-        Box::new(VaultProvider::new(
+    pub fn build_provider(&self) -> Result<VariablesProvider> {
+        Ok(Box::new(VaultProvider::new(
             &self.url,
             &self.token,
             &self.mount,
             self.prefix.as_deref(),
-        ))
+        )))
     }
 }
 
 #[derive(Debug, Default, Deserialize)]
 #[serde(deny_unknown_fields)]
 pub struct AzureKeyVaultVariablesProviderOpts {
-    pub client_id: String,
-    pub client_secret: String,
-    pub tenant_id: String,
     pub vault_url: String,
-    #[serde(default)]
-    pub authority_host: AzureAuthorityHost,
+    pub client_id: Option<String>,
+    pub client_secret: Option<String>,
+    pub tenant_id: Option<String>,
+    pub authority_host: Option<AzureAuthorityHost>,
 }
 
 impl AzureKeyVaultVariablesProviderOpts {
-    pub fn build_provider(&self) -> VariablesProvider {
-        Box::new(AzureKeyVaultProvider::new(
-            &self.client_id,
-            &self.client_secret,
-            &self.tenant_id,
+    pub fn build_provider(&self) -> Result<VariablesProvider> {
+        let auth_config_runtime_vars = [&self.client_id, &self.tenant_id, &self.client_secret];
+        let any_some = auth_config_runtime_vars.iter().any(|&var| var.is_some());
+        let any_none = auth_config_runtime_vars.iter().any(|&var| var.is_none());
+
+        if any_none && any_some {
+            // some of the service principal auth options were specified, but not enough to authenticate.
+            return Err(anyhow!("Azure Key Vault provider requires each client_id, client_secret, and tenant_id to be provided. If none are provided, the provider will attempt to authenticate using ambient authentication (e.g. env vars, Azure CLI, Managed Identity, Workload Identity"));
```

Comment:
I had a go at wordsmithing this to focus more on "what is the problem and what is the remedy."  I'm not super happy with the result but sharing it in the hope that it will inspire you to come up with the perfect wording...

```suggestion
            return Err(anyhow!("The current runtime config specifies some but not all of the Azure KeyVault 'client_id', 'client_secret', and 'tenant_id' values. Provide the missing values to authenticate to Azure KeyVault with the given service principal, or remove all these values to authenticate using ambient authentication (e.g. env vars, Azure CLI, Managed Identity, Workload Identity)."));
```

Reply:
Yours is an improvement on what I had. I think it is clear, but might benefit from a link to auth docs when they are authored. I don't think any brief error message is going to provide the clarity needed to understand all of the options implied.